RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+
+This issue to some extent is also present in ERC-4337.
+
+As an alternative, if Native Account Abstraction is to be implemented in coordination with EIP-3540,
+relying on the concept of "code sections" it introduces is a superior approach.
+
+## Specification
+
+### Validation and PostTransaction code sections
+
+The EOF format describes a separation of 'code' and 'data' sections,
+marked with `kind_code (0x02)` and `kind_data (0x04)` kind markers.
+
+The contract that may have a role in an Account Abstraction transaction, either as a Sender, a Paymaster or a Deployer,
+has to contain a section marked with the following kind marker:
+
+```
+kind_abstraction = 0x05
+```
+
+This section is equivalent to a code section.
+
+Its code can be executed during a regular transaction execution and has no special effects.
+If it is the first code section of a contract, it can act as an entry point during regular transaction execution.
+
+Only a single section of `kind_abstraction` is allowed in a contract.
+This rule is validated during contract creation.
+
+### Execution entry point for Account Abstraction transaction type participant entity (Sender, Paymaster and Deployer)
+
+During a regular contract code execution, its behaviour is defined as follows by EIP-3540:
+```
+Execution starts at the first byte of code section 0, and pc is set to 0
+```
+
+However, if a contract is referenced in an `AA_TX_TYPE` transaction as a Sender, Paymaster or a Deployer,
+execution starts at the first byte of code section with `kind_abstraction` marker, and `pc` is set to 0.
+
+If the specified contract does not contain such a section, or is not an EOF contract, the transaction is not valid.
+
+### Encoding inputs for different execution frames
+
+We define the following "action" identifiers:
+
+```
+sender_deployment = 0x01
+sender_validation = 0x02
+paymaster_validation = 0x03
+paymaster_posttx = 0x04
+```
+
+#### Sender Deployment
+
+Inputs to the `Deployer` contract are not defined by the protocol and are controlled by the `deployerData` parameter.
+
+The sender deployment frame MUST result in the `sender` address becoming initialized with contract code.
+
+#### Sender Validation
+
+Inputs to the `Sender` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
+```
+
+The sender validation frame MUST return 32 bytes `validationData` that is interpreted as:
+
+```solidity
+abi.encodePacked(MAGIC_VALUE_SENDER, validUntil, validAfter)
+```
+
+#### Paymaster Validation
+
+Inputs to the `Paymaster` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
+```
+
+The frame must return a bytes array that is interpreted as:
+
+```solidity
+abi.encode(context, MAGIC_VALUE_PAYMASTER, validUntil, validAfter)
```

Comment:
See comment above re ABI and validity time range.  With different sections we don't need the magic value, the validity time range can be verified at the end of validation via an ERC-7562 change that seems safe to me (but let's make sure).  That leaves us with only `context`.  Since it's now just one value, we don't need to abi.encode it.  Just return the context.

Reply:
paymaster return value should be reversed, so it doesn't require ABI-encoding: it can return `encodePacked(validFrom,validAfter,context)`
(that is: 6 bytes from, 6-bytes after, <len-12> bytes context), taking advantage of the fact we have exactly one dynamic-len param. it will get ugly to encode more than one.
(again, I agree that with EOF, we don't need magic_value anymore)